cpp/include/raft/linalg/unary_op.cuh

```diff
@@ -116,32 +117,60 @@ void unary_op(const raft::handle_t& handle, InType in, OutType out, Lambda op)
 }
 
 /**
- * @brief perform element-wise binary operation on the input arrays
- * This function does not read from the input
- * @tparam InType Input Type raft::device_mdspan
+ * @brief Perform an element-wise unary operation on the input index into the output array
+ *
+ * To be deprecated. Please use the index_unary_op instead.
+ *
+ * @tparam OutType Output Type raft::device_mdspan
  * @tparam Lambda the device-lambda performing the actual operation
  * @param[in] handle raft::handle_t
- * @param[inout] in Input/Output
+ * @param[out] out Output
  * @param[in] op the device-lambda
  * @note Lambda must be a functor with the following signature:
- *       `InType func(const InType& val);`
+ *       `void func(out_value_t* out_location, index_t idx);`
  */
-template <typename InType, typename Lambda, typename = raft::enable_if_output_device_mdspan<InType>>
-void write_only_unary_op(const raft::handle_t& handle, InType in, Lambda op)
+template <typename OutType,
+          typename Lambda,
+          typename = raft::enable_if_output_device_mdspan<OutType>>
+void write_only_unary_op(const raft::handle_t& handle, OutType out, Lambda op)
 {
-  RAFT_EXPECTS(raft::is_row_or_column_major(in), "Input must be contiguous");
+  RAFT_EXPECTS(raft::is_row_or_column_major(out), "Output must be contiguous");
 
-  using in_value_t = typename InType::value_type;
+  using out_value_t = typename OutType::value_type;
 
-  if (in.size() <= std::numeric_limits<std::uint32_t>::max()) {
-    writeOnlyUnaryOp<in_value_t, Lambda, std::uint32_t>(
-      in.data_handle(), in.size(), op, handle.get_stream());
+  if (out.size() <= std::numeric_limits<std::uint32_t>::max()) {
+    writeOnlyUnaryOp<out_value_t, Lambda, std::uint32_t>(
+      out.data_handle(), out.size(), op, handle.get_stream());
   } else {
-    writeOnlyUnaryOp<in_value_t, Lambda, std::uint64_t>(
-      in.data_handle(), in.size(), op, handle.get_stream());
+    writeOnlyUnaryOp<out_value_t, Lambda, std::uint64_t>(
+      out.data_handle(), out.size(), op, handle.get_stream());
   }
 }
 
+/**
+ * @brief Perform an element-wise unary operation on the input index into the output array
+ *
+ * @tparam OutType Output mdspan type
+ * @tparam Lambda  The unary operation type
+ * @param[in]  handle The raft handle
+ * @param[out] out    Output array
+ * @param[in]  op     The unary operation
+ * @note Lambda must be a functor with the following signature:
+ *       `OutT func(const IdxT& idx);`
+ */
+template <typename OutType,
+          typename Lambda,
+          typename = raft::enable_if_output_device_mdspan<OutType>>
+void index_unary_op(const raft::handle_t& handle, OutType out, Lambda op)
```

Comment:
I think we should revise the naming and find something more appropriate.
1. As long as we provide one flat index into the lambda, I think we either should narrow down the output type to a rank-1 mdspans (vectors only), or indicate that we provide offsets, and not indices.
2. Analogously to `unary_op`, I'd expect `xxx_unary_op` also accept one input array. Would you consider something more explicit, like `generate_with_offset` and `generate_with_index` or `generate_indexed`?
3. In the face of mdspan api, I'd really better go with the `offset` in the name, so that later we can add another functions which would provide a proper multi-dimensional indexing.

Reply:
I would slightly prefer `generate_from_offset` or maybe `map_offset`.